### PR TITLE
Issue 1259/1344: Views

### LINKED ISSUE #1259: Migration: Views
**Description**
Migration of Views panel.

**Tasks**
Create a new Views component

**Extras**
design: https://xd.adobe.com/view/8a6b23f1-36b6-4419-5fe3-78a60d7ee17d-fc5a/screen/ee904147-e7a4-4eb1-93d1-c809c00dcf5d/Views

### PR DESCRIPTION
**Pull request by simplypixi**
This fixes #1259 and #1344.

### CODE REVIEW

**Comment by simplypixi** [comment]
@sanmont3drepo , @carmenfan I removed re-loading previous model, but I noticed two bugs, which appear on staging too:

1. When I load Lego Architecture model, go to Teamspaces and then load Lego Federation model the viewer throws some errors related to the pins. It can't to remove the pins, because they exists in pins map but not in unity context.
image

2. When I go to Teamspace while the model is loading, the viewer doesn't cancel request/model rendering, so it's appears in the Teamspaces background. I think Pavol reported exactly the same bug, but it seems to be an issue on viewer side (e.g $destroy in viewer.component never calls).

What do you think about it?

**Comment by carmenfan** [comment]
> @sanmont3drepo , @carmenfan I removed re-loading previous model, but I noticed two bugs, which appear on staging too:
> 
> 1. When I load Lego Architecture model, go to Teamspaces and then load Lego Federation model the viewer throws some errors related to the pins. It can't to remove the pins, because they exists in pins map but not in unity context.

I'm guessing this only happens on initialisation? It's probably worth taking a look during issue migration, as the pins should be consistent with unity provided the model has started loading. I think I've seen this in the pass and it doesn't really affect the functionality as the pins were loaded after the model started loading.

> 2. When I go to Teamspace while the model is loading, the viewer doesn't cancel request/model rendering, so it's appears in the Teamspaces background. I think Pavol reported exactly the same bug, but it seems to be an issue on viewer side (e.g $destroy in viewer.component never calls).
> 
> What do you think about it?

This is a known issue. It's not to do with the component not getting destroyed. It's API requests that has been fired in unity, and before they come back reset has been called. So the whole dataset has been wiped in Unity, but then we get the callback.

We know this happens for some time now and if we were too address it we need to change i in Unity, so I wouldn't worry too much about it. (as long as it's clearing the viewer if you exit after the model has been loaded

**Comment by simplypixi** [comment]
@carmenfan @sanmont3drepo @dleszcz Rebase and ready for review ;)

File: frontend/components/index.ts [hunk 1]
```diff
@@ -60,6 +60,4 @@ import './utils/js/utils.filter';
 import './viewer/js/multi-select.service';
 import './viewer/js/viewer.component';
 import './viewer/js/viewer.service';
-import './viewpoints/js/viewpoints.component';
```
Comment (sanmont3drepo) [inline frontend/components/index.ts]: Should delete the viewpoints angular component folder also.

File: frontend/modules/viewer/viewer.sagas.ts [hunk 2]
```diff
@@ -30,7 +31,7 @@ export function* waitForViewer() {
 		const ViewerService = yield getAngularService('ViewerService') as any;
 		yield ViewerService.initialised.promise;
 	} catch (error) {
-		console.error(error);
+		yield put(DialogActions.showErrorDialog('initialise', 'viewer', error));
```
Comment (sanmont3drepo) [inline frontend/modules/viewer/viewer.sagas.ts]: - If the error is an enpoint error should use the DialogActions.showEndpointErrorDialog action instead. If it is a string its ok. Please check every call.

File: frontend/routes/viewer/components/views/components/viewItem/viewItem.component.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	viewpoint: any;
+	active: boolean;
+	editMode: boolean;
+	teamspace: string;
+	modelId: string;
+	onCancelEditMode: () => void;
+	onSaveEdit: (values) => void;
+	onDelete?: (event) => void;
+	onOpenEditMode?: () => void;
+	handleClick?: (viewpoint) => void;
```
Comment (sanmont3drepo) [inline frontend/routes/viewer/components/views/components/viewItem/viewItem.component.tsx]: - Can we rename **handleClick** to **onClick** ?

Comment (simplypixi) [reply frontend/routes/viewer/components/views/components/viewItem/viewItem.component.tsx]: Yes, definitely. I don't know why it was named like this. :/

File: frontend/components/index.ts (hunk 1 above)
Comment (simplypixi) [reply frontend/components/index.ts]: done

**Comment by simplypixi** [comment]
> I think there is a bug with the resizing of the card.
> When you have an empty list of views the height is small and stays small until you refresh the page.
> 
> Now:
> resizing bad
> Empty list:
> resizing bad2
> 
> How it should look:
> resizing ok

Fixed ;)

**Comment by sanmont3drepo** [comment]
I think everything is ok, can I merge this @carmenfan ?

**Comment by carmenfan** [comment]
Let me put this through the @paloknapo test first :)

**Comment by carmenfan** [comment]
@simplypixi 
@paloknapo says everything works, but the icons should be in one row:
image
image

Also the icons should be in grey, like this:
image

**Comment by simplypixi** [comment]
> @simplypixi
> @paloknapo says everything works, but the icons should be in one row:
> image
> image
> 
> Also the icons should be in grey, like this:
> image

I changed these icons on Risks branch about a week ago. I think this PR was created before the views got new icons' look. Move this fix here? ;)

**Comment by simplypixi** [comment]
> > @simplypixi
> > @paloknapo says everything works, but the icons should be in one row:
> > image
> > image
> > Also the icons should be in grey, like this:
> > image
> 
> I changed these icons on Risks branch about a week ago. I think this PR was created before the views got new icons' look. Move this fix here? ;)

Fixed ;)

**Comment by simplypixi** [comment]
> @simplypixi
> 
> On the Angular side, when there's a deleted group (or closed issues / resolved risks) happened on another client and you get the update via the chat service, We get this little message before the item disappears. Can we do the same for views?
> 
> 

Sure. ;) Should it look the same?

**Comment by simplypixi** [comment]
@carmenfan Fixed and rebased ;)

**Comment by carmenfan** [comment]
@simplypixi looking good :)

But this seems to happen still?

> Selected view should unselect when you add a new view:
>

**Comment by simplypixi** [comment]
> @simplypixi looking good :)
> 
> But this seems to happen still?
> 
> > Selected view should unselect when you add a new view:
> > 

Fixed ;)

**Pull request merged**